프리코스 이후 재도전!!

kyw0716 opened the pull request.

Commit:
docs: 기능 목록 및 예외 상황 추가

Commit:
feat: 다리 길이 입력 기능

Commit:
feat: 이동할 칸 선택 값 입력 기능

Commit:
docs: 기능 목록 추가 - 정답 다리 생성 기능

Commit:
feat: 정답 다리 생성 기능

Commit:
feat: 정답 다리와 입력된 값 비교 기능

Commit:
feat: 현재 위치 출력 기능

Commit:
feat: 최종 게임 결과 출력 기능

Commit:
feat: 입력 값 유효성 검사 기능

Commit:
refactor: 상수 처리

jjjjjeongwon commented on src/BridgeGame.js:
```diff
@@ -1,20 +1,57 @@
+      size,
+      BridgeRandomNumberGenerator.generate
+    );
+    console.log(this.#answer);
```
엇 여기 `console.log()`가 남아있네요!
1주차 피드백 내용 중에 '불필요한 console.log를 남기지 않는다'라는 내용이 있어서 확인해보시면 좋을 것 같아요:)
[1주차 피드백](https://docs.google.com/document/d/1fxW5IrWeJ6VATWJqE9F5NDiRifrqLRsChXFVicVEQdw/edit#heading=h.o7t80t1h9qjc)

kyw0716 replied:
오오 감사합니다! 확인한다고 추가했던걸 깜빡하고 못지웠나보네요 ㅠㅠ

jjjjjeongwon commented on src/BridgeGame.js:
```diff
@@ -1,20 +1,57 @@
-  retry() {}
+  retry() {
+    this.#currentPosition = 0;
+    this.#totalTrial++;
```
저도 놓쳤던 부분인데, 에어비앤비 컨벤션에 따르면 `++`보다는`+= 1`을 사용할 것을 제안하고 있더라구요!
확인해보시면 좋을 것 같아요 :)
[Airbnb JS Style 13.6](https://github.com/airbnb/javascript#variables--unary-increment-decrement)

kyw0716 replied:
이 부분은 처음 알았네요 많이 배워갑니다 감사합니다!!

jjjjjeongwon commented on src/BridgeMaker.js:
```diff
@@ -7,7 +7,18 @@ const BridgeMaker = {
+  makeBridge(size, generateRandomNumber) {
+    const answerBridge = [];
+
+    for (let i = 0; i < size; i++) {
+      const randomNumber = generateRandomNumber();
+
+      if (parseInt(randomNumber) === 0) answerBridge.push("D");
+      if (parseInt(randomNumber) === 1) answerBridge.push("U");
+    }
```
```suggestion
    for (let i = 0; i < size; i++) {
      answerBridge.push( generateRandomNumber() ? "U" : "D");
    }
```
코드 상에서 0은 false로, 1은 true로 인식한다고 해요.
이 부분을 활용해서 if문 대신 삼항연산자를 활용해서 이런 방식으로 중복코드를 줄여보는건 어떨까요 ?:)

jjjjjeongwon replied:
"U"와 "D"는 의미를 가지고 있는 값이니 상수 처리를 해서 가지고 와서 사용하는게 좋을 것 같습니다 !

kyw0716 replied:
상수 처리를 다 했다고 생각했는데 남는 부분이 있었네요! 코드를 완성하고 상수 처리를 하지 않고 코드를 작성하면서 상수 처리를 해야겠어요!! 감사합니당

jjjjjeongwon reviewed (commented):
코드 잘 보고 갑니다 :) 
MVC패턴 적용에 많은 시간을 들여 고민하신게 느껴지네요!

Commit:
refactor: 피어 리뷰 사항 적용

- 불필요한 console.log 제거
- ++ => += 1로 변경
- 의미 있는 값 상수 처리
- 1,0을 true false로 인식하는 것을 이용해 if문 수정

mayreel0 commented on src/controller/BridgeGameController.js:
```diff
@@ -0,0 +1,66 @@
+  }
+
+  progressGame(moving) {
+    if (this.#gameModel.getCurrentMovingCorrect(moving)) {
+      if (this.#gameModel.getIsLastPosition())
+        return this.printResult(ResultState.GAME_SUCCESSED);
+      this.#gameModel.move();
+
+      return this.inputMoving();
+    }
+    return this.inputRestart();
```
이렇게 하면 deps를 줄일 수 있을 것 같습니다
```suggestion
    if (!this.#gameModel.getCurrentMovingCorrect(moving)) {
      return this.inputRestart();
    }
    if (this.#gameModel.getIsLastPosition()) {
      return this.printResult(ResultState.GAME_SUCCESSED);
    }
    this.#gameModel.move();

    return this.inputMoving();
```

mayreel0 commented on src/views/OutputView.js:
```diff
@@ -0,0 +1,84 @@
+    let wrongMap = [];
+
+    if (map[map.length - 1] === BridgeDirection.UP)
+      [...wrongMap] = this.makeUpWrongMap(upMap, downMap);
```
이 둘에는 어떤 차이가 있을까요?
```js
[...wrongMap] = this.makeUpWrongMap(upMap, downMap);
```
```js
wrongMap = this.makeUpWrongMap(upMap, downMap);
```

mayreel0 commented on src/BridgeMaker.js:
```diff
@@ -7,7 +9,21 @@ const BridgeMaker = {
+    for (let i = 0; i < size; i++) {
+      const randomNumber = generateRandomNumber();
+
+      if (parseInt(randomNumber)) {
```
10진수로 변환하기 위해 parseInt를 사용한다면 기본값이 10진수가 아니기 때문에 반드시 두번째 인자를 10으로 지정해주어야합니다. 
https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Global_Objects/parseInt#radix

mayreel0 commented on src/BridgeMaker.js:
```diff
@@ -7,7 +9,21 @@ const BridgeMaker = {
+  makeBridge(size, generateRandomNumber) {
+    const answerBridge = [];
+
+    for (let i = 0; i < size; i++) {
+      const randomNumber = generateRandomNumber();
+
+      if (parseInt(randomNumber)) {
+        answerBridge.push(BridgeDirection.UP);
+        continue;
+      }
+      answerBridge.push(BridgeDirection.DOWN);
+    }
```
for문을 사용하지 않고 명시적으로 배열을 생성할 수 있는 메서드나 방법이 있습니다.
https://tesseractjh.tistory.com/103

mayreel0 commented on src/utils/InputValidator.js:
```diff
@@ -0,0 +1,32 @@
+  },
+
+  validateMoving(input) {
+    const movingCommand = [BridgeDirection.UP, BridgeDirection.DOWN];
```
배열을 새로 생성할 필요 없이 Object 메서드를 사용해도 될 것 같습니다.
```js
const movingCommand = Object.values(BridgeDirection);
```

mayreel0 commented on src/views/OutputView.js:
```diff
@@ -0,0 +1,84 @@
+  },
+
+  makeUpWrongMap(upMap, downMap) {
+    downMap.pop();
+    downMap.push(MoveState.FAILED);
+    upMap.pop();
+    upMap.push(MoveState.NOTHING);
+    return [upMap, downMap];
```
https://evan-moon.github.io/2020/01/05/what-is-immutable/

hwanfront reviewed (commented):
LGTM
